### PR TITLE
Initial implementations of the TPU Vortexor

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem
Framework for Solana TPU vortexor which allows core validator to offload TPU receiving, sigverify and deduplications.

#### Summary of Changes
This is the first installment of the change. This is the framework part. It starts the TPU and forward stream based on the CLI configurations. In future PRs we will have the sigverify, forwarder, subscription management. The code has no impact to the existing validator.

Added a unit test for starting the vortexor, send some transactions and stops it.

Fixes #

### CODE REVIEW

File: vortexor/src/cli.rs
Comment (sakridge) [inline vortexor/src/cli.rs:95]: why does it need dynamic port range?

File: vortexor/Cargo.toml
Comment (sakridge) [inline vortexor/Cargo.toml:57]: crate should be a bin right?

File: vortexor/src/cli.rs
Comment (lijunwangs) [reply vortexor/src/cli.rs:95]: This is similar to how it is done in the validator side which allows the user to specify the range for the tpu addresses.

File: vortexor/Cargo.toml
Comment (lijunwangs) [reply vortexor/Cargo.toml:57]: Yes. The crate generate a bin solana-vortexor. It can also support generating a library which can be used by other crates to develop. For example, I see these

ls target/debug | grep vortex
libsolana_vortexor.d
libsolana_vortexor.rlib
solana-vortexor
solana-vortexor.d

File: streamer/src/nonblocking/testing_utilities.rs
Comment (lijunwangs) [inline streamer/src/nonblocking/testing_utilities.rs:189]: This is moved from nonblocking::quic::test to here to be reused for other testing

**Comment by behzadnouri** [comment]
Is there a design document or anything for this?
like what is the motivation and end goal here? how is the validator going to talk with this new binary? how does the binary obtain stakes to do stake-based QoS etc? does the binary need access to the private key of the validator? etc ...

**Comment by lijunwangs** [comment]
> Is there a design document or anything for this? like what is the motivation and end goal here? how is the validator going to talk with this new binary? how does the binary obtain stakes to do stake-based QoS etc? does the binary need access to the private key of the validator? etc ...

The motivation is to make TPU streaming, sig verification more scalable. The readme captures this: https://raw.githubusercontent.com/anza-xyz/agave/11908fbbcfab7194f954177c2da0646249ee308a/vortexor/Readme.md. I have set up a review meeting to go over this to give more context. The vortexor will retrieve the stake map via regular RPC service from configured validator -- the RPC can be either pairing validator or any other trusted RPC validator in the network supporting the RPC pub/sub service. The Vortexor does not need the private key to the validator.

**Comment by behzadnouri** [comment]
> The Vortexor does not need the private key to the validator.

don't we need the private key for QUIC handshake ?

**Comment by lijunwangs** [comment]
> > The Vortexor does not need the private key to the validator.
> 
> don't we need the private key for QUIC handshake ?

A private key is used to sign the TLS certificate. But the client does not care about if the key is the same as the one of the validator advertising the TPU address (The vortexor address). The client code does not check that and we have no use case to enforce that check -- this is the same currently using --public-tpu-address. Without requiring using the same private key making it more secure and simplify deployment.

**Comment by behzadnouri** [comment]
> A private key is used to sign the TLS certificate. But the client does not care about if the key is the same as the one of the validator advertising the TPU address (The vortexor address). The client code does not check that and we have no use case to enforce that check

How does the client prune its connection cache when it becomes too big?
Isn't that by pubkey => stake mapping? and isn't the pubkey from QUIC handshake?

**Comment by lijunwangs** [comment]
> > A private key is used to sign the TLS certificate. But the client does not care about if the key is the same as the one of the validator advertising the TPU address (The vortexor address). The client code does not check that and we have no use case to enforce that check
> 
> How does the client prune its connection cache when it becomes too big? Isn't that by pubkey => stake mapping? and isn't the pubkey from QUIC handshake?

We randomly select a connection to prune -- the current connection cache is agnostic of the stake map.

**Comment by lijunwangs** [comment]
@pgarg66 @behzadnouri @bw-solana @sakridge -- can you follow up on this PR request? I have updated the readme.md with some of feedback I got in the table top review. Thanks

**Comment by pgarg66** [comment]
Hey @lijunwangs, could you post the comments/summary from the table-top review? It'll be good to have them captured here in this PR. 

Also, there were some action items based on the review. It'll be great to have tracking issues for those, so that reviewers know what's coming down the line for Vortexor.

**Comment by bw-solana** [comment]
I don't have a ton of feedback on the code for now, but a few comments (most of which are just reiterating what came up during live discussion):

- I think it would be good to map out how Jito relayers and DZ networks might interface with the vortexor
- It would be good to call out if vortexor and core validator can be run on the same box
- It would be good to provide some details on how operators can upgrade from current validator to vortexor + core validator configuration.
- It would be good to specify details on how core validator ingest might work. Is this connection provided all of the bandwidth by default? Can core validator also ingest from other sources? How is throttling handled when dealing with multiple sources of ingress (e.g. multiple vortexors and public TPU connections)?

**Comment by lijunwangs** [comment]
> I don't have a ton of feedback on the code for now, but a few comments (most of which are just reiterating what came up during live discussion):
> 
> * I think it would be good to map out how Jito relayers and DZ networks might interface with the vortexor
> * It would be good to call out if vortexor and core validator can be run on the same box
> * It would be good to provide some details on how operators can upgrade from current validator to vortexor + core validator configuration.
> * It would be good to specify details on how core validator ingest might work. Is this connection provided all of the bandwidth by default? Can core validator also ingest from other sources? How is throttling handled when dealing with multiple sources of ingress (e.g. multiple vortexors and public TPU connections)?

Thanks for the comments. Updated the Readme.md with this.

**Comment by lijunwangs** [comment]
> Hey @lijunwangs, could you post the comments/summary from the table-top review? It'll be good to have them captured here in this PR.
> 
> Also, there were some action items based on the review. It'll be great to have tracking issues for those, so that reviewers know what's coming down the line for Vortexor.

I have put the discussion point we had into the Readme.md -- as I think it is a design questions. I have created a project to track these efforts:

https://github.com/orgs/anza-xyz/projects/15/views/1

File: vortexor/src/vortexor.rs
Comment (pgarg66) [inline vortexor/src/vortexor.rs]: This will collide with the metrics name for the TPU streamer in main validator. Can we rename it?

File: vortexor/src/vortexor.rs
Comment (pgarg66) [inline vortexor/src/vortexor.rs]: Why not merge the `new` with `create_vortexor`? We are creating the config object in this function which gets unraveled in `new`. It'll save on that.

File: vortexor/Readme.md
Comment (bw-solana) [inline vortexor/Readme.md]: FYI - looks like this diagram formatting got messed up: https://github.com/lijunwangs/solana/tree/tpu_vortexor/vortexor

File: vortexor/src/vortexor.rs
Comment (pgarg66) [inline vortexor/src/vortexor.rs]: It's confusing as a user of vortexor that we are summing up staked and unstaked count on the forwarding streamer.
Should we add another CLI arg specific for the max connections on forwarding streamer?

File: vortexor/src/vortexor.rs
Comment (pgarg66) [inline vortexor/src/vortexor.rs:52]: Is this functionally different than what we do in the main validator client? Wondering if the code can be refactored and used here.

File: vortexor/src/vortexor.rs
Comment (pgarg66) [inline vortexor/src/vortexor.rs]: Should we move these two const to cli.rs and rename to add `DEFAULT`? Currently it's giving an impression that these are the hard limits.

**Review by pgarg66 — COMMENTED** [review]
Some suggestions/comments.

In its current form it looks like a wrapper on top of part of existing QUIC and TPU functionality. The follow on PRs would add more specifics for Vortexor.

It looks in the right direction to me, but I would like another person to review it as well.

File: vortexor/Readme.md
Comment (lijunwangs) [reply vortexor/Readme.md]: Will fix it up

File: vortexor/src/vortexor.rs
Comment (lijunwangs) [reply vortexor/src/vortexor.rs]: We can -- I was torn between using the quic_streamer_tpu and having a new metric name. There is pro and cons. The pro sharing the name is to it is easier to look at the network wide metrics -- after all they are ingesting transactions the same as regular validator and metrics are relevant. The con is it is harder to look at vortexor specific metrics.

File: vortexor/src/vortexor.rs
Comment (lijunwangs) [reply vortexor/src/vortexor.rs]: Will fix

File: vortexor/src/vortexor.rs
Comment (lijunwangs) [reply vortexor/src/vortexor.rs]: Will do

File: vortexor/src/vortexor.rs
Comment (pgarg66) [reply vortexor/src/vortexor.rs]: The bigger con in when there's a hybrid network, where only some nodes are using vortexor. It'll be hard to differentiate between the metrics.

Comment (lijunwangs) [reply vortexor/src/vortexor.rs]: agreed

File: vortexor/src/vortexor.rs
Comment (lijunwangs) [reply vortexor/src/vortexor.rs:52]: In the first pr it does is just a thin wrapper of the streamer. I thought about refactoring it and decided against it as  1. I see little code change than wrapping the arguments and sending to streamer.2. The vortexor  will be loaded with other functionality such as key updater interaction, sigverify interaction, forwarder interactions.

File: vortexor/Readme.md
Comment (lijunwangs) [reply vortexor/Readme.md]: Fixed. Looks at least need to start with 4 white spaces to format correctly

File: vortexor/src/vortexor.rs
Comment (lijunwangs) [reply vortexor/src/vortexor.rs]: done

**Review by pgarg66 — APPROVED** [review]
Thanks for addressing the comments! It looks good to me.
I would suggest to get another review approval before merging this.

**Comment by lijunwangs** [comment]
> Thanks for addressing the comments! It looks good to me. I would suggest to get another review approval before merging this.

Checked with Brennan, I will address any additional feedback in follow-on PRs. Merging it to form a baseline.

File: vortexor/Cargo.toml
Comment (steviez) [inline vortexor/Cargo.toml:6]: I see the crate has `publish = false` set now, but from what I can tell, it looks like this would be a separate service run on a separate bin. Should this live in a different repo ?

Comment (bw-solana) [reply vortexor/Cargo.toml:6]: I asked @lijunwangs this same question. His current thinking is that developing in the monorepo will get something out faster but that it would make sense to split this into a separate repo at some point.

It's not clear to me what the level of urgency is w/ the vortexor. Splitting out after the fact is likely harder than doing it up front, but I understand this could increase the work to get Vortexor v0 out.

More generally, long term, I see multiple ways of ingesting packets with different assumptions about "quality" that could be made:

- public via TPU
- Jito bundles
- Double Zero network
- Vortexor

Coding different solutions and/or having elevated status for a potential source seems wrong. Splitting vortexor into separate repo probably makes it easier to do the right thing here

Comment (steviez) [reply vortexor/Cargo.toml:6]: Gotcha, that seems reasonable. As long as it stays with `publish = false` (which also means that no published crates depend on it), I'm fine with it being in the monorepo. Just thinking about making sure we don't expand our surface in consideration of stuff like https://github.com/anza-xyz/agave/issues/3724

Comment (lijunwangs) [reply vortexor/Cargo.toml:6]: The reason for keeping  them in the same repo is there are dependencies between the vortexor and other components in agave validator which requires some refactoring such as streamer, sigverify and etc. Having them in the beginning will get us the implementation quicker by reducing the time waiting for the dependency to be published. I agreed that after we get stable first version of the vortexor it is right thing to move it separate repo. We did something similar before for the Postgres geyser plugin implementation.

**Pull request merged**
